element_array_ephys/ephys_precluster.py

```diff
@@ -0,0 +1,896 @@
+
+            self.insert1(dict(key,
+                              lfp_sampling_rate=oe_probe.lfp_meta['sample_rate'],
+                              lfp_time_stamps=lfp_timestamps,
+                              lfp_mean=lfp.mean(axis=0)))
+
+            electrode_query = (probe.ProbeType.Electrode
+                               * probe.ElectrodeConfig.Electrode
+                               * EphysRecording & key)
+            probe_electrodes = {key['electrode']: key
+                                for key in electrode_query.fetch('KEY')}
+
+            for channel_idx in np.array(oe_probe.lfp_meta['channels_ids'])[lfp_channel_ind]:
+                electrode_keys.append(probe_electrodes[channel_idx])
+        else:
+            raise NotImplementedError(f'LFP extraction from acquisition software'
+                                      f' of type {acq_software} is not yet implemented')
+
+        # single insert in loop to mitigate potential memory issue
+        for electrode_key, lfp_trace in zip(electrode_keys, lfp):
+            self.Electrode.insert1({**key, **electrode_key, 'lfp': lfp_trace})
+
+
+# ------------ Clustering --------------
+
+@schema
+class ClusteringMethod(dj.Lookup):
+    definition = """
+    # Method for clustering
+    clustering_method: varchar(16)
+    ---
+    clustering_method_desc: varchar(1000)
+    """
+
+    contents = [('kilosort', 'kilosort clustering method'),
+                ('kilosort2', 'kilosort2 clustering method')]
+
+
+@schema
+class ClusteringParamSet(dj.Lookup):
+    definition = """
+    # Parameter set to be used in a clustering procedure
+    paramset_idx:  smallint
+    ---
+    -> ClusteringMethod    
+    paramset_desc: varchar(128)
+    param_set_hash: uuid
+    unique index (param_set_hash)
+    params: longblob  # dictionary of all applicable parameters
+    """
+
+    @classmethod
+    def insert_new_params(cls, processing_method: str, paramset_idx: int,
+                          paramset_desc: str, params: dict):
+        param_dict = {'clustering_method': processing_method,
+                      'paramset_idx': paramset_idx,
+                      'paramset_desc': paramset_desc,
+                      'params': params,
+                      'param_set_hash':  dict_to_uuid(params)}
+        param_query = cls & {'param_set_hash': param_dict['param_set_hash']}
+
+        if param_query:  # If the specified param-set already exists
+            existing_paramset_idx = param_query.fetch1('paramset_idx')
+            if existing_paramset_idx == paramset_idx:  # If the existing set has the same paramset_idx: job done
+                return
+            else:  # If not same name: human error, trying to add the same paramset with different name
+                raise dj.DataJointError(
+                    'The specified param-set'
+                    ' already exists - paramset_idx: {}'.format(existing_paramset_idx))
+        else:
+            cls.insert1(param_dict)
```

Comment:
```suggestion
        cls.insert1(param_dict)
```

DataJoint already handles these duplication errors better than this code.